### DISCUSSION TITLE
Slow movement/lagging when using rdk plugin with multiple apertures.

### DISCUSSION THREAD

**Discussion by jordandeakin** [jspsych/jsPsych#978] [Q&A] [answered]
Hi,

I am using the rdk plugin to display varying numbers of random dot kinematograms which will be part of an online experiment. I use the plugin to display trials consisting of 5, 10 or 15 apertures. When 5 apertures are presented, the dots move smoothly and at the correct speed. However, with trials containing 10 and 15 apertures, the dots move slower and the movement is not smooth. 

Strangely, this doesn't seem to happen when running the experiment in Safari browser. The RDKs in all trials (regardless of number of apertures) move as expected and the speed of the dots does not change depending on number of apertures. With Firefox, Chrome and Microsoft Edge however, the problem gets progressively worse the more apertures are added. 

This is how I display the apertures. I am not sure if this is a problem with the browser, the computer used or my code, I can provide the code used to make the timeline if this helps.

```
var trialRDKs= {
type: 'rdk',
coherence: jsPsych.timelineVariable("coherence"),
correct_choice: jsPsych.timelineVariable("correct_choice"),
coherent_direction: jsPsych.timelineVariable("coherent_direction"),
number_of_apertures: jsPsych.timelineVariable("number_of_apertures"),
aperture_width: 90,
aperture_height: 90,
number_of_dots: 50,
choices: ['z','m'],
aperture_center_x: jsPsych.timelineVariable("posX"),
aperture_center_y: jsPsych.timelineVariable("posY"),
response_ends_trial: true,
background_color: 'black',
dot_color: 'white',
trial_duration: 100000,
data: {practice: 1}
};


var procedure= {
  timeline: [fixation,trialRDKs,feedback],
  timeline_variables: trialSequence,
};
```

Any help would be greatly appreciated!
Thank you :)

**Comment by jodeleeuw** [comment]
Hi @jordandeakin ,

I suspect that you are simply hitting the performance limit of the browser and machine. It may be that Safari has a more efficient rendering pipeline for this particular kind of display change. I don't think there's anything you could change at the level of specifying the individual trials that will fix this. 

It may be that there are changes to the plugin code that would make it more efficient, but I don't know. @vrsivananda is the plugin's creator. He may have some ideas. 

I expect you might see some trade off with the number of dots. Perhaps 10 apertures with 10 dots each is no worse than 2 apertures with 50 dots each? This may be a clue about what parts of the pipeline are lagging.

**Reply by jordandeakin** [reply]
Okay I see, thank you very much!

**Comment by vrsivananda** [comment, marked as answer]
Hi @jordandeakin, 

I think @jodeleeuw is correct in that you are likely hitting the performance limit of the browser/computer. After a certain number of dots the RDK starts to slow down noticeably. In my testing before publishing the code, Chrome seemed to perform the best, so it is interesting the Safari is doing better for you.

I'm sorry I'm not exactly sure how to fix it (other than reducing the number of dots per aperture), but if you are right that Safari is handling it well, you could potentially make sure that your subjects are using Safari in their browser. Hopefully this doesn't bias your sample in a way that affects your study. [Here](https://www.geeksforgeeks.org/how-to-detect-the-user-browser-safari-chrome-ie-firefox-and-opera-using-javascript/) is a resource on how to implement that, if you need it.

Another way around it is to check post-experiment how the stimuli was displayed. E.g. `frame_rate`, `number_of_frames`, and `frame_rate_array` data should be able to give you an idea of which participants had better browsers/machines that could display it as intended.

Hope this helps!

**Reply by jordandeakin** [reply]
That is very helpful, I just wanted to make sure it wasn't something I was doing to make it less efficient. Thank you very much for your help!
